lib/elixir/src/elixir_expand.erl

```diff
@@ -728,6 +729,30 @@ allowed_in_context(_, _Arity, guard) ->
 allowed_in_context(_, _, _) ->
   true.
 
+maybe_warn_struct_comparison({{'.', _, [erlang, Op]}, Meta, [Left, Right]}, E)
+    when Op =:= '>'; Op =:= '<'; Op =:= '=<'; Op =:= '>=' ->
+  Result =
+    case struct_expression(Left) of
+      false -> struct_expression(Right);
+      Struct -> Struct
+    end,
+
+  case Result of
+    false -> ok;
+    _ -> elixir_errors:form_warn(Meta, ?key(E, file), ?MODULE, {struct_comparison, Result})
```

Comment:
currently the warning is:

```elixir
iex(1)> Date.utc_today() >= %Date{year: 2018, month: 1, day: 1}
warning: invalid comparison with struct literal Elixir.Date. Comparison operators (>, <, >=, <=) perform structural and not semantic comparison. Comparing with a struct literal is unlikely to give a meaningful result. Modules typically define a compare/2 function that can be used for semantic comparison
  iex:1

true
```

I'd be happy to send a patch with `elixir_aliases:inspect(Result)` to at least make it to `(...) with struct literal Date. (...)`.

With `'Elixir.Macro':to_string(Right)` (or Left) we could turn it to `(...) with struct literal %Date{calendar: Calendar.ISO, year: 2018, month: 1, day: 1}`. Maybe it would make sense to go all the way back to sigil?

Reply:
by trial and error :) I got to:

> warning: invalid comparison with struct literal ~D[2018-01-01]

with:

```erlang
      {M, _} = 'Elixir.Code':eval_quoted(Right),
      R = 'Elixir.Kernel':inspect(M),
```

not familar with internals; if it's worth pursuing what would be appropriate APIs to call here?

Reply:
Let's go with `elixir_aliases:inspect(Result)` and an updated test. Thanks!

Reply:
The second one seems a bit risky. :)